### PR TITLE
[SPMD] add FQN argument to Override.replacement

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100473

Differential Revision: [D45486089](https://our.internmc.facebook.com/intern/diff/D45486089)

### CODE REVIEW

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45486089).

File: test/distributed/_spmd/test_tracing.py [hunk 1]
```diff
@@ -516,7 +516,9 @@ class AssertOverride(Override):
             def __init__(self, outer):
                 self.outer = outer
 
-            def replacement(self, orig_submodule: torch.nn.Module) -> torch.nn.Module:
+            def replacement(
+                self, orig_submodule: torch.nn.Module, fqn: str
```
Comment (wanchaol) [inline test/distributed/_spmd/test_tracing.py:520]: nit: should `fqn` be the first argument instead of `orig_submodule`?

**Review by wanchaol — APPROVED** [review]
Curious what's the motivation of this patch, is the same instance of `origin_submodule` being shared by multiple fqns, and we only want to switch one fqn pointed submodule?

File: test/distributed/_spmd/test_tracing.py (hunk 1 above)
Comment (mrshenli) [reply test/distributed/_spmd/test_tracing.py:520]: good point, let me update.

**Comment by mrshenli** [comment]
> Curious what's the motivation of this patch, is the same instance of origin_submodule being shared by multiple fqns, and we only want to switch one fqn pointed submodule?

@yifuwang mentioned a case on the sparse side where there are multiple submodules of the same type sharing the same global sharding plan. We need to know which part of the plan each submodule corresponds to. So having the module instance along is not sufficient, and we use fqn to distinguish them.

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45486089).

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45486089).

**Pull request closed without merging**